src/shader/shaderUtils.js

```diff
@@ -234,4 +234,42 @@ export const shaderUtils = /* glsl */`
 		return ( - 0.69813170079773212 * x * x - 0.87266462599716477 ) * x + 1.5707963267948966;
 
 	}
+
+	bool intersectsRectangle( vec3 position, vec3 normal, vec3 u, vec3 v, vec3 rayOrigin, vec3 rayDirection, out float dist ) {
+
+		float dt = dot( rayDirection, normal );
+		float t = ( dot( normal, position ) - dot( normal, rayOrigin ) ) / dt;
+
+		if ( t > EPSILON ) {
+
+			vec3 hit = rayOrigin + rayDirection * t;
+			vec3 vi = hit - position;
+			float a1 = dot( u, vi );
+			if ( a1 >= - 0.5 && a1 <= 0.5 ) {
+
+				float a2 = dot( v, vi );
+				if ( a2 >= - 0.5 && a2 <= 0.5 ) {
+
+					dist = t;
+					return true;
+
+				}
+
+			}
+
+		}
+
+		return false;
+
+	}
```

Comment:
The computation of `t` can be rewritten as follows, saving a `dot` product.

```glsl
float t = dot(position - rayOrigin, normal) / dot(rayDirection, normal);
```

I've pushed a commit to the spot lights PR that includes this change, it could be factored out into a separate function if it ends up repeated more.

Reply:
I don't really have a reference, but it basically computes the intersection of the ray with the plane as defined by the rect area light (through the `u` and `v` vectors) and then checks if that point falls within the width and height from the center position of the light. I've added a comment describing the process and included a reference for the first part ([ray-plane intersection](https://lousodrome.net/blog/light/2020/07/03/intersection-of-a-ray-and-a-plane/)).

@richardassar Thanks for the optimization. The reference I've added for the ray-plane intersection also mentions this optimization.